bots/machine/testvm_core.py

```diff
@@ -0,0 +1,557 @@
+#!/usr/bin/python3 -u
+# -*- coding: utf-8 -*-
+
+# This file is part of Cockpit.
+#
+# Copyright (C) 2013 Red Hat, Inc.
+#
+# Cockpit is free software; you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation; either version 2.1 of the License, or
+# (at your option) any later version.
+#
+# Cockpit is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+
+
+import os
+import time
+import socket
+import signal
+import subprocess
+import tempfile
+import select
+import errno
+import sys
+
+BOTS_DIR = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
+TEST_DIR = os.path.join(os.path.dirname(BOTS_DIR), "test")
+DEFAULT_IDENTITY_FILE = os.path.join(BOTS_DIR, "machine", "identity")
+
+ROOT_USER_TUPLE = ["root", "foobar"]
+ADMIN_USER_TUPLE = ["admin", "foobar"]
+TEST_USER_TUPLE = ["test", "superhardpasswordtest5554"]
+
+class Timeout:
+    """ Add a timeout to an operation
+        Specify machine to ensure that a machine's ssh operations are canceled when the timer expires.
+    """
+    def __init__(self, seconds=1, error_message='Timeout', machine=None):
+        if signal.getsignal(signal.SIGALRM) != signal.SIG_DFL:
+            # there is already a different Timeout active
+            self.seconds = None
+            return
+
+        self.seconds = seconds
+        self.error_message = error_message
+        self.machine = machine
+
+    def handle_timeout(self, signum, frame):
+        if self.machine:
+            if self.machine.ssh_process:
+                self.machine.ssh_process.terminate()
+            self.machine.disconnect()
+
+        raise RuntimeError(self.error_message)
+
+    def __enter__(self):
+        if self.seconds:
+            signal.signal(signal.SIGALRM, self.handle_timeout)
+            signal.alarm(self.seconds)
+
+    def __exit__(self, type, value, traceback):
+        if self.seconds:
+            signal.alarm(0)
+            signal.signal(signal.SIGALRM, signal.SIG_DFL)
+
+
+class Failure(Exception):
+    def __init__(self, msg):
+        self.msg = msg
+
+    def __str__(self):
+        return self.msg
+
+
+class RepeatableFailure(Failure):
+    pass
```

Comment:
These two errors are exclusively being used in testvm.py, they shouldn't be moved here.

Reply:
I wanted to not put them here, but they were used by ssh connection methods like: https://github.com/cockpit-project/cockpit/pull/10486/files#diff-e426e1a754f7ec0103f2f51e49de9452R169